### PR TITLE
SR-12486: `T.self is Any.Protocol` is broken

### PR DESCRIPTION
**Pull request by tbkka**
This turned out to be fallout from https://github.com/apple/swift/pull/27572
which was in turn motivated by our confusing metatype syntax when generic variables are bound to protocols.

In particular, the earlier PR was an attempt to make the expression
`x is T.Type` (where `T` is a generic type variable bound to a protocol `P`)
behave the same as
`x is P.Type` (where `P` is a protocol).
Unfortunately, the generic `T.Type` actually binds to `P.Protocol` in this case (not `P.Type`), so the original motivation was flawed, and as it happens, `x is T.Type` already behaved the same as `x is P.Protocol` in this situation.

This PR reverts that earlier change and beefs up some of the tests around these behaviors.

Resolves SR-12486

Resolves rdar://62201613

Reverts PR #27572

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

**Review by jckarter — APPROVED** [review]
Thanks @tbkka !

File: test/Interpreter/generic_casts.swift [hunk 1]
```diff
@@ -215,135 +218,150 @@ func genericAnyAsUnconditional<T>(type: Any.Type, to: T.Type, expected: Bool) ->
...
-print(genericAnyIs(type: P.self, to: P.self, expected: true)) // CHECK: true
-print(nongenericAnyIsPConforming(type: PS.self)) // CHECK: true
-print(genericAnyIs(type: PS.self, to: P.self, expected: true)) // CHECK-ONONE: true
-print(nongenericAnyIsPConforming(type: PE.self)) // CHECK: true
-print(genericAnyIs(type: PE.self, to: P.self, expected: true)) // CHECK-ONONE: true
-print(nongenericAnyIsPConforming(type: PC.self)) // CHECK: true
-print(genericAnyIs(type: PC.self, to: P.self, expected: true)) // CHECK-ONONE: true
-print(nongenericAnyIsPConforming(type: PCSub.self)) // CHECK: true
-print(genericAnyIs(type: PCSub.self, to: P.self, expected: true)) // CHECK-ONONE: true
+print(#line, nongenericAnyIsPType(type: P.self)) // CHECK: [[@LINE]] false
```
Comment (mikeash) [inline test/Interpreter/generic_casts.swift:221]: Good move adding all these `#line`s. I hope you were able to automate that.

**Review by mikeash — APPROVED** [review]

File: test/Interpreter/generic_casts.swift (hunk 1 above)
Comment (tbkka) [reply test/Interpreter/generic_casts.swift:221]: Emacs macros FTW!

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
Testing again after PR #32664

**Comment by tbkka** [comment]
@swift-ci Please test

**Pull request merged**